File: Chapter 03/Java/src/main/java/com/grokkingsoftwarearchitecture/chapter03/section_3_3_1_srp/after/PlayerRepository.java
package com.grokkingsoftwarearchitecture.chapter03.section_3_3_1_srp.after;

/**
 * Responsibility 3: Data Persistence.
 * * ARCHITECTURE NOTE: This class handles all database infrastructure. If we decide 
 * to migrate from a SQL database to a NoSQL database, or if our ORM framework 
 * changes, this is the *only* class that needs to be updated. The Player and 
 * Tactics classes remain completely untouched.
 */
public class PlayerRepository {
    
    /**
     * Persists the player's statistics to the underlying data store.
     * * @param player The player whose stats need saving.
     */
    public void saveStats(Player player) {
        System.out.println("  [Database] Saving " + player.name + "'s game stats to the database.");
    }
}

File: Chapter 03/Python/section_3_3_1_srp/after/player_repository.py
from .player import Player
"""
Responsibility 3: Data Persistence (The Repository Pattern).
* ARCHITECTURE NOTE: This class handles all infrastructure concerns. It acts 
as a mediator between our domain (Player) and the data mapping layer. 
* Because this logic is isolated, you can change your storage backend from 
an SQL database to a NoSQL store, or even a simple CSV file for local 
testing, by only modifying this repository. The core game logic remains 
completely untouched.
"""
class PlayerRepository:
    def save_stats(self, player: 'Player'):
        """
        Persists the player's statistics to the underlying data store.
        """
        print(f"  [Database] Saving {player.name}'s game stats to the database.")